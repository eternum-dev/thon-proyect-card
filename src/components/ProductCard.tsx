import React, { createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import { InitialValue, OnChangeArgs, Product, ProductCartHandlers, ProductContextProps } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    children: (args: ProductCartHandlers) => React.JSX.Element;
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: OnChangeArgs) => void;
    value?: number;
    initialValues?: InitialValue;
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

    const { counter, increaseBy, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter: counter,
            increaseBy: increaseBy,
            product: product,
            maxCount: initialValues?.maxCount
        }}>
            <div className={className ? styles.productCard + ' ' + className : styles.productCard}
                style={style}
            >
                {children({
                    count: counter,
                    maxCount: initialValues?.maxCount,
                    reset,
                    increaseBy,
                    isMaxCountReached,
                    product
                })}
            </div>
        </Provider>
    )
}