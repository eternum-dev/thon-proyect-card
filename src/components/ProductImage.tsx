import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import noimage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';


export interface Props {
    img?: string,
    className?: string,
    style?: React.CSSProperties
}

export const ProductImage = ({ img, className, style }: Props) => {

    let imgToShow: string;
    const { product } = useContext(ProductContext);

    img ? imgToShow = img : product.img ?
        imgToShow = product.img
        : imgToShow = noimage;

    return (
        <img
            className={` ${styles.productImg} ${className}`}
            src={imgToShow}
            alt="Product image"
            style={style}
        />
    )
}
