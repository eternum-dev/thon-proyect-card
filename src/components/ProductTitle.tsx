import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';



export interface Props {
    title?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductTitle = ({ title, className, style }: Props) => {

    let titleToShow: string;
    const { product } = useContext(ProductContext);

    title ?
        titleToShow = title :
        titleToShow = product.title

    return (
        <span
            className={`${styles.productDescription} ${className}`}
            style={style}
        >
            {titleToShow}
        </span>
    )
}
