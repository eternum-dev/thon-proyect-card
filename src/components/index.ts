import { ProductCard as ProductCardHoc } from './ProductCard';

import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductTitle } from './ProductTitle';
import { ProductCardHocProps } from '../interfaces/interfaces';


export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';


export const ProductCard: ProductCardHocProps = Object.assign(ProductCardHoc, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});


export default ProductCard;