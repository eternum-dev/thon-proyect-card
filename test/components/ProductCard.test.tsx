import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product2 } from '../data/products';


describe('Pruebas en el componente <ProductCard />', () => {

    const { act } = renderer;

    test('Debe de motrar el componente correctamente', () => {

        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <h1>produc card</h1>
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON).toMatchSnapshot();
    });

    test('Debe incrementar el contador', () => {
        const wrapper = renderer.create(

            <ProductCard product={product2}>
                {
                    ({ count, increaseBy }) => (
                        <>
                            <h1>produc card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}></button>
                        </>
                    )
                }
            </ProductCard>
        )
        let tree = wrapper.toJSON();
        act(() => {
            (tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe('1')

    });
})