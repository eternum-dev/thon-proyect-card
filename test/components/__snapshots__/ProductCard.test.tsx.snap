// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <ProductCard /> Debe de motrar el componente correctamente 1`] = `[Function]`;
